### ISSUE TITLE
Models + Parameters

### ISSUE THREAD

**Issue by oliver-pola** [oliver-pola/OneMillionPostsCorpus#2]
Bevor wir alle möglichen Varianten des Models und der Parameter jeweils in eigenen Commits raushauen, dachte ich es wäre besser einfach mal hier ein paar Versuchsergebnisse reinzustellen...

**Comment by oliver-pola** [comment]
Ziemlich viel Dropout:
```
        model = tf.keras.Sequential()
        model.add(tf.keras.Input(shape=(padded_length, embedding_dim)))
        model.add(tf.keras.layers.SpatialDropout1D(0.5))
        model.add(tf.keras.layers.LSTM(lstm_out, dropout=0.4, recurrent_dropout=0.4))
        model.add(tf.keras.layers.Dense(128, activation=tf.nn.relu))
        model.add(tf.keras.layers.Dropout(0.1))
        model.add(tf.keras.layers.Dense(1, activation=tf.keras.activations.sigmoid))
        model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
```
Wird dann ziemlich langsam, aber dadurch zieht das Training nicht so davon und Validation kann eher mithalten:
```
Epoch 1/200
3239/3239 - 15s - loss: 0.6971 - accuracy: 0.5387 - val_loss: 0.6873 - val_accuracy: 0.5167
Epoch 21/200
3239/3239 - 10s - loss: 0.6690 - accuracy: 0.5437 - val_loss: 0.7118 - val_accuracy: 0.4833
Epoch 26/200
3239/3239 - 9s - loss: 0.6576 - accuracy: 0.5659 - val_loss: 0.7107 - val_accuracy: 0.5944
Epoch 100/200
3239/3239 - 9s - loss: 0.5428 - accuracy: 0.7092 - val_loss: 0.7752 - val_accuracy: 0.6028
Epoch 149/200
3239/3239 - 9s - loss: 0.4817 - accuracy: 0.7617 - val_loss: 0.8123 - val_accuracy: 0.6250
Epoch 192/200
3239/3239 - 9s - loss: 0.4502 - accuracy: 0.7777 - val_loss: 0.8133 - val_accuracy: 0.6417
Epoch 195/200
3239/3239 - 9s - loss: 0.4307 - accuracy: 0.7882 - val_loss: 0.8241 - val_accuracy: 0.6333
Epoch 200/200
3239/3239 - 9s - loss: 0.4340 - accuracy: 0.7879 - val_loss: 0.8175 - val_accuracy: 0.6306
```
Sieht aus als könnte man durchaus noch weiter trainieren, obwohl der Validtion-Loss nie den gewünschten Bogen von abwärts nach aufwärts macht, sondern von Anfang an gleich aufwärts geht.

**Comment by oliver-pola** [comment]
Mit dem Dropout wie zuvor, habe mich mal versucht die größe der Layer zu verdoppeln / halbieren. Für `lstm_out` konnte ich keine Verbesserung erreichen, aber beim Reduzieren des Dense Layer von 128 auf 32. 
Dann wollte ich noch mit der Learning Rate spielen. Hatten wir bisher nicht angegeben, aber der Default für den Adam Optimizer ist wohl `learning_rate=0.001`. Irgendwie scheint es mit höherer Rate besser zu klappen (obwohl ich vermutet hätte bei kleiner Rate kommt man zum selben Ergebnis, es dauert einfach nur länger ?!?):
```
lstm_out = 128
model = tf.keras.Sequential()
model.add(tf.keras.Input(shape=(padded_length, embedding_dim)))
model.add(tf.keras.layers.SpatialDropout1D(0.5))
model.add(tf.keras.layers.LSTM(lstm_out, dropout=0.4, recurrent_dropout=0.4))
model.add(tf.keras.layers.Dense(32, activation=tf.nn.relu))
model.add(tf.keras.layers.Dropout(0.1))
model.add(tf.keras.layers.Dense(1, activation=tf.keras.activations.sigmoid))
model.compile(optimizer='adam', learning_rate=0.01, loss='binary_crossentropy', metrics=['accuracy'])
```
```
Epoch 1/200
3239/3239 - 20s - loss: 0.6975 - accuracy: 0.5215 - val_loss: 0.6947 - val_accuracy: 0.5167
Epoch 44/200
3239/3239 - 14s - loss: 0.6252 - accuracy: 0.5847 - val_loss: 0.7530 - val_accuracy: 0.5917
Epoch 57/200
3239/3239 - 14s - loss: 0.5913 - accuracy: 0.6548 - val_loss: 0.7579 - val_accuracy: 0.6417
Epoch 66/200
3239/3239 - 14s - loss: 0.5752 - accuracy: 0.6730 - val_loss: 0.7376 - val_accuracy: 0.6250
Epoch 85/200
3239/3239 - 14s - loss: 0.5301 - accuracy: 0.7215 - val_loss: 0.7632 - val_accuracy: 0.6472
Epoch 100/200
3239/3239 - 14s - loss: 0.5060 - accuracy: 0.7394 - val_loss: 0.7841 - val_accuracy: 0.6222
Epoch 150/200
3239/3239 - 13s - loss: 0.4264 - accuracy: 0.7956 - val_loss: 0.8953 - val_accuracy: 0.6028
Epoch 154/200
3239/3239 - 14s - loss: 0.4272 - accuracy: 0.7935 - val_loss: 0.8934 - val_accuracy: 0.6222
Epoch 198/200
3239/3239 - 13s - loss: 0.3644 - accuracy: 0.8391 - val_loss: 0.9681 - val_accuracy: 0.6389
Epoch 199/200
3239/3239 - 14s - loss: 0.3739 - accuracy: 0.8253 - val_loss: 0.9512 - val_accuracy: 0.6083
Epoch 200/200
3239/3239 - 13s - loss: 0.3790 - accuracy: 0.8219 - val_loss: 0.9780 - val_accuracy: 0.6194
```
https://imgur.com/Jm1ihOp

**Comment by oliver-pola** [comment]
Ich war auch mal neugierig, wie es bei anderen Categories so aussieht. Die Ausgabe `train labels mean = 0.46` bzw. `val labels mean = 0.48` gibt übrigens den Anteil der Daten mit Category = 1 an. Gleiche Parameter wie https://github.com/oliver-pola/OneMillionPostsCorpus/issues/2#issuecomment-581365454 ansonsten Commit f2166bd. Loss + Accuracy jeweils nach 200 Epochen:

```
SentimentNegative
labels mean = 0.47
loss: 0.3790 - accuracy: 0.8219 - val_loss: 0.9780 - val_accuracy: 0.6194

SentimentNeutral
labels mean = 0.52
loss: 0.3814 - accuracy: 0.8228 - val_loss: 0.9828 - val_accuracy: 0.5778

SentimentPositive
labels mean = 0.01
loss: 0.0096 - accuracy: 0.9975 - val_loss: 0.1356 - val_accuracy: 0.9833

OffTopic
labels mean = 0.15
loss: 0.2352 - accuracy: 0.9061 - val_loss: 0.4176 - val_accuracy: 0.8722

Inappropriate
labels mean = 0.08
loss: 0.1291 - accuracy: 0.9481 - val_loss: 0.4104 - val_accuracy: 0.8944

Discriminating
labels mean = 0.08
loss: 0.1229 - accuracy: 0.9521 - val_loss: 0.5383 - val_accuracy: 0.8972

PossiblyFeedback
labels mean = 0.02
loss: 0.0144 - accuracy: 0.9948 - val_loss: 0.2375 - val_accuracy: 0.9722

PersonalStories
labels mean = 0.01
loss: 0.0107 - accuracy: 0.9948 - val_loss: 0.1078 - val_accuracy: 0.9861

ArgumentsUsed
labels mean = 0.28
loss: 0.1993 - accuracy: 0.9154 - val_loss: 0.6332 - val_accuracy: 0.7722
```
Ich weiß nicht so genau, was ich davon halten soll, dass nur 1% der Daten auf SentimentPositive = 1 gesetzt sind. Zumindest ist das in Training und Validation gleich und es wird zu 99% bzw. 98% korrekt klassifiziert.

Hier mal alle Ergebnisse, aus dem neuen Ordner "output" (ignoriert von git):
[output.zip](https://github.com/oliver-pola/OneMillionPostsCorpus/files/4148542/output.zip)

**Comment by oliver-pola** [comment]
Inzwischen habe ich eine Meinung zu den dünn verteilten Daten: Wenn nur 1% der Posts SentimentPositive=1 (bzw. PersonalStories=1) sind, dann könnte man natürlich einfach konstant 0 predicten und käme auf accuracy=0.99. Alles was weniger ist, ist schlecht. Auch die 0.98 hört sich eigentlich erst einmal ganz gut an, ist aber schlecht.
Mathematisch: ok ist, wenn val_accuracy > (1 - val_labels_mean)

**Comment by oliver-pola** [comment]
Ich habe mal Precision und Recall sowie eine Confusion-Matrix eingebaut und dann noch mal jeweils 100 Epochen trainieren lassen.

confusion matrix = [[true neg, false pos], [false neg, true pos]]

```
SentimentNegative
Epoch 100/100
3239/3239 - 17s - loss: 0.5265 - accuracy: 0.7178 - precision: 0.7149 - recall: 0.6632 - val_loss: 0.7533 - val_accuracy: 0.6139 - val_precision: 0.5762 - val_recall: 0.7076
confusion matrix = [[100, 89], [50, 121]]

SentimentNeutral
Epoch 100/100
3239/3239 - 15s - loss: 0.5437 - accuracy: 0.7061 - precision: 0.6983 - recall: 0.7576 - val_loss: 0.8174 - val_accuracy: 0.6278 - val_precision: 0.6562 - val_recall: 0.6495
confusion matrix = [[100, 66], [68, 126]]

SentimentPositive
Epoch 100/100
3239/3239 - 15s - loss: 0.0176 - accuracy: 0.9932 - precision: 0.9412 - recall: 0.4324 - val_loss: 0.0937 - val_accuracy: 0.9861 - val_precision: 1.0000 - val_recall: 0.1667
confusion matrix = [[354, 0], [5, 1]]

OffTopic
Epoch 100/100
3239/3239 - 15s - loss: 0.3035 - accuracy: 0.8833 - precision: 0.7872 - recall: 0.3606 - val_loss: 0.5187 - val_accuracy: 0.8028 - val_precision: 0.3333 - val_recall: 0.0597
confusion matrix = [[285, 8], [63, 4]]

Inappropriate
Epoch 100/100
3239/3239 - 12s - loss: 0.2002 - accuracy: 0.9290 - precision: 0.7312 - recall: 0.2491 - val_loss: 0.3121 - val_accuracy: 0.9222 - val_precision: 0.7500 - val_recall: 0.1000
confusion matrix = [[329, 1], [27, 3]]

Discriminating
Epoch 100/100
3239/3239 - 12s - loss: 0.2110 - accuracy: 0.9358 - precision: 0.9474 - recall: 0.1488 - val_loss: 0.4322 - val_accuracy: 0.8861 - val_precision: 0.3333 - val_recall: 0.0250
confusion matrix = [[318, 2], [39, 1]]

PossiblyFeedback
Epoch 100/100
3239/3239 - 12s - loss: 0.0422 - accuracy: 0.9870 - precision: 0.7500 - recall: 0.4839 - val_loss: 0.1738 - val_accuracy: 0.9694 - val_precision: 0.3333 - val_recall: 0.1000
confusion matrix = [[348, 2], [9, 1]]

PersonalStories
Epoch 100/100
3239/3239 - 12s - loss: 0.0310 - accuracy: 0.9910 - precision: 0.9375 - recall: 0.3488 - val_loss: 0.0576 - val_accuracy: 0.9889 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00
confusion matrix = [[356, 0], [4, 0]]

ArgumentsUsed
Epoch 100/100
3239/3239 - 16s - loss: 0.2603 - accuracy: 0.8855 - precision: 0.8327 - recall: 0.7445 - val_loss: 0.5051 - val_accuracy: 0.7889 - val_precision: 0.6630 - val_recall: 0.5755
confusion matrix = [[223, 31], [45, 61]]
```

Bei den Categories mit wenig 1en bekommen wir natürlich auch sehr wenig true positives (also precision und recall sind gering) und mehr false negatives als false positives, also die Tendenz einfach immer nur 0 auszugeben ist schon deutlich zu sehen.

Hier wieder die kompletten Logs und die Grafiken zu Precision und Recall:
[output_100.zip](https://github.com/oliver-pola/OneMillionPostsCorpus/files/4209958/output_100.zip)

**Comment by oliver-pola** [comment]
Habe noch mal alle durchlaufen lassen, jetzt mit
- All
- Bidirectional für Single
- Callbacks für Single

Details könnt ihr den Logs und Images entnehmen.
Zip: https://easyupload.io/3iwdz8
